### PR TITLE
[Impeller] Emplace directly into host buffer (avoid VBB) for text data

### PR DESCRIPTION
**Pull request by jonahwilliams**
From local testing, this shaves off about 0.3-4 ms of pure allocation overhead from https://github.com/flutter/flutter/issues/127760

### Before

image

### After

image

### CODE REVIEW

File: impeller/core/host_buffer.h [hunk 1]
```diff
@@ -95,6 +95,22 @@ class HostBuffer final : public std::enable_shared_from_this<HostBuffer>,
...
+  /// @brief      Emplace non-uniform data (like contiguous vertices) onto the
+  ///             host buffer directly. The buffer ptr is guaranteed to have at
+  ///             least enough storage for the length of data.
+  ///
+  /// @param[in]  cb            A callback that will be passed a ptr to the
+  ///                           underlying host buffer.
+  ///
+  /// @return     The buffer view.
+  ///
+  BufferView EmplaceCallback(size_t length,
```
Comment (chinmaygarde) [inline impeller/core/host_buffer.h:110]: Nit: I'd rather just take advantage of function overloading and use the name `Emplace`.

File: impeller/core/host_buffer.h [hunk 2]
```diff
@@ -95,6 +95,22 @@ class HostBuffer final : public std::enable_shared_from_this<HostBuffer>,
                                    size_t length,
                                    size_t align);
 
+  using EmplaceProc = std::function<void(uint8_t* buffer)>;
+
+  //----------------------------------------------------------------------------
+  /// @brief      Emplace non-uniform data (like contiguous vertices) onto the
```
Comment (chinmaygarde) [inline impeller/core/host_buffer.h:101]: "Emplaces undefined data onto the managed buffer and gives the caller a chance to update it using the specified callback. The buffer is guaranteed to have enough space for length bytes. It is the responsibility of the caller to not exceed the bounds of the buffer returned in the EmplaceProc.".

File: impeller/core/host_buffer.cc [hunk 3]
```diff
@@ -56,6 +56,19 @@ BufferView HostBuffer::Emplace(const void* buffer, size_t length) {
...
 
+BufferView HostBuffer::EmplaceCallback(size_t length,
+                                       size_t align,
+                                       const EmplaceProc& cb) {
+  auto old_length = GetLength();
+  if (!Truncate(old_length + length)) {
+    return {};
+  }
+  generation_++;
+  cb(GetBuffer() + old_length);
```
Comment (chinmaygarde) [inline impeller/core/host_buffer.cc:67]: Emplace and EmplaceCallback can be dried up by making Emplace call EmplaceCallback with a callback that does the `memmove`.

File: impeller/core/host_buffer.h (hunk 1 above)
Comment (chinmaygarde) [reply impeller/core/host_buffer.h:110]: I suspect we will only use this sparingly for performance but its good to have the open as needed.

**Review by chinmaygarde — APPROVED** [review]
Some nits but lgtm.

File: impeller/core/host_buffer.cc (hunk 3 above)
Comment (chinmaygarde) [reply impeller/core/host_buffer.cc:67]: Also, let's check that CB is valid. If not, emplacing undefined data is fine.

Comment (dnfield) [reply impeller/core/host_buffer.cc:67]: Alternatively consider making this return a `std::pair`

Comment (dnfield) [reply impeller/core/host_buffer.cc:67]: e.g. `std::pair<BufferView, uintptr_t> Reserve(size_t length, size_t align);` and then on this line you 

```cpp
return std::make_pair(BufferView{...}, GetBuffer() + old_length);
```

Comment (dnfield) [reply impeller/core/host_buffer.cc:67]: Emplace could still be dried up as Chinmay is suggesting to internally call `Reserve` and use the pointer.

**Comment by dnfield** [comment]
LGTM, stylistic comments that are not meant to be blocking.

**Pull request merged**